velox/functions/sparksql/RegexFunctions.cpp

```diff
@@ -135,79 +124,76 @@ struct RegexpReplaceFunction {
     if (utf8Position > stringInput.size() + 1) {
-      result.resize(inputStringPiece.size());
-      std::memcpy(
-          result.data(), inputStringPiece.data(), inputStringPiece.size());
+      result = stringInput;
       return;
     }
 
-    re2::StringPiece prefix(inputStringPiece.data(), utf8Position);
-    re2::StringPiece targetStringPiece(
-        inputStringPiece.data() + utf8Position,
-        inputStringPiece.size() - utf8Position);
+    re2::RE2* patternRegex = getRegex(pattern.str());
+    re2::StringPiece replaceStringPiece = toStringPiece(replace);
 
+    std::string prefix(stringInput.data(), utf8Position);
     std::string targetString(
-        targetStringPiece.data(), targetStringPiece.size());
-    RE2::GlobalReplace(&targetString, *patternRegex, replaceStringPiece);
+        stringInput.data() + utf8Position, stringInput.size() - utf8Position);
 
-    if (targetString.size() || prefix.size()) {
-      result.resize(prefix.size() + targetString.size());
-      std::memcpy(result.data(), prefix.data(), prefix.size());
-      std::memcpy(
-          result.data() + prefix.size(),
-          targetString.data(),
-          targetString.size());
-    } else {
-      result.resize(0);
-    }
+    RE2::GlobalReplace(&targetString, *patternRegex, replaceStringPiece);
+    result = prefix + targetString;
   }
 
  private:
-  re2::RE2* getCachedRegex(const std::string& pattern) const {
-    auto it = patternCache_.find(pattern);
-    if (it != patternCache_.end()) {
+  re2::RE2* getRegex(const std::string& pattern) const {
+    auto it = cache_.find(pattern);
+    if (it != cache_.end()) {
       return it->second.get();
     }
     VELOX_USER_CHECK_LT(
-        patternCache_.size(),
+        cache_.size(),
         kMaxCompiledRegexes,
         "regexp_replace hit the maximum number of unique regexes: {}",
         kMaxCompiledRegexes);
     checkForCompatiblePattern(pattern, "regexp_replace");
     auto patternRegex = std::make_unique<re2::RE2>(pattern);
     auto* rawPatternRegex = patternRegex.get();
     checkForBadPattern(*rawPatternRegex);
-    patternCache_.emplace(pattern, std::move(patternRegex));
+    cache_.emplace(pattern, std::move(patternRegex));
     return rawPatternRegex;
   }
 
-  mutable folly::F14FastMap<std::string, std::unique_ptr<re2::RE2>>
-      patternCache_;
+  template <typename TString>
+  size_t getUTFLength(TString& input, size_t maxCharacters) {
+    const char* str = input.data();
+    size_t bytes = input.size();
+    size_t utf8Position = 0;
+    size_t numCodePoints = 0;
+    while (numCodePoints < maxCharacters && utf8Position < bytes) {
+      int charLength = utf8proc_char_length(str + utf8Position);
+      VELOX_USER_CHECK_GT(
+          charLength, 0, "regexp_replace encountered invalid UTF-8 character");
+      ++numCodePoints;
+      utf8Position += charLength;
+    }
+    return utf8Position;
+  }
+
+  mutable folly::F14FastMap<std::string, std::unique_ptr<re2::RE2>> cache_;
```

Comment:
Does this need to be 'mutable'?

Reply:
Tested locally and removed mutable. Got the following compile error:


`/velox/velox/functions/sparksql/RegexFunctions.cpp:176:5: error: cannot convert ‘const folly::F14FastMap<std::__cxx11::basic_string<char>, std::unique_ptr<re2::RE2> >*’ to ‘folly::f14::detail::F14BasicMap<folly::f14::detail::VectorContainerPolicy<std::__cxx11::basic_string<char>, std::unique_ptr<re2::RE2>, void, void, void, std::integral_constant<bool, true> > >* 
176 |     cache_.emplace(pattern, std::move(patternRegex));`

The mutable keyword is still new to me but after looking online, I believe this error occured because getRegex was a constant function, so to modify the member "cache_" of RegexpReplaceFunction within it, cache_ needed to be mutable.

 I don't recall why I made it constant (it could have been to resolve the inverse of this error), and I don't know the implications of removing it. That said everything compiles and all tests pass when you remove mutable from cache_ and const from getRegex. If there's nothing setting off alarms in your head with these changes, they will be included in the next push later tonight.

Reply:
> remove mutable from cache_ and const from getRegex

Let's do that. Thanks.